Feature: Golang Http Tcp Bridge (#36667)
envoyproxy/envoy: docs/root/configuration/http/tcp_bridge/golang.rst

.. _config_http_tcp_bridge_golang:

Golang Http Tcp Bridge
======================

> This bridge enables an Http client to connect to a Tcp server via a Golang plugin, facilitating **Protocol Convert** from Http to any Rpc protocol in Envoy.

Reviewer: Rpc -> RPC

Author: done

> Notice: the bridge is designed for sync data flow between go and c, so **DO NOT** use goroutine in go side.

Reviewer: it's not exactly.

people could create new goroutines, but should not touch the request in these goroutines, they could be background goroutines.

Author: got it~

   ...
   go version goX.YZ linux/amd64

For example, to build the ``.so`` for a ``foo`` plugin, you might run:

.. code-block:: console

   $ bazel run @go_sdk//:bin/go build -- --buildmode=c-shared  -v -o path/to/output/libfoo.so path/to/src/foo

Configuration
-------------

.. tip::
   This filter should be configured with the type URL
   :ref:`type.googleapis.com/envoy.extensions.upstreams.http.tcp.golang.v3alpha.Config <envoy_v3_api_msg_extensions.upstreams.http.tcp.golang.v3alpha.Config>`.

A prebuilt Golang Http Tcp Bridge  ``my_plugin.so`` might be configured as follows:

.. literalinclude:: /_configs/go/golang-http-tcp-bridge.yaml
   :language: yaml
   :linenos:
   :lines: 38-44
   :lineno-start: 39
   :emphasize-lines: 2-7
   :caption: :download:`golang-http-tcp-bridge.yaml </_configs/go/golang-http-tcp-bridge.yaml>`

Extensible plugin configuration
-------------------------------

Envoy's Go plugins can specify and use their own configuration.

Below is a very simple example of how such a plugin might be configured in Envoy:

.. literalinclude:: /_configs/go/golang-http-tcp-bridge-with-config.yaml
   :language: yaml
   :linenos:

>    :lines: 38-50

Reviewer: seems the line number is incorrect? see: https://storage.googleapis.com/envoy-pr/a4ef3f6/docs/configuration/http/tcp_bridge/golang.html

Author: yes，have optimize it.

.. _config_http_tcp_bridge_golang:

> Golang Http Tcp Bridge

Reviewer 2: ```suggestion
Golang HTTP TCP Bridge
```

======================

> This bridge enables an Http client to connect to a Tcp server via a Golang plugin, facilitating **Protocol Convert** from Http to any RPC protocol in Envoy.

Reviewer 2: ```suggestion
This bridge enables an HTTP client to connect to a TCP server via a Golang plugin, facilitating **Protocol Convert** from HTTP to any RPC protocol in Envoy.
```

Author: fix tit, thks~

> Note that the bridge is designed for sync-data-flow between go and c, so when you create new goroutines, **DO NOT** touch the request in these goroutines, they could be background goroutines.

Reviewer 2: ```suggestion
.. note::
  The bridge is designed for sync-data-flow between go and c, so when you create new goroutines, **DO NOT** touch the request in these goroutines, they could be background goroutines.
```

Author: fix tit, thks~

The Bridge allows `Golang <https://go.dev/>`_ to be run during both the request
and response flows of upstream.

Go plugins used by this bridge can be recompiled independently of Envoy.

See the `Envoy's Golang HTTP-TCP Bridge proposal
<https://github.com/envoyproxy/envoy/issues/35749>`_
for more details on the filter's implementation.

Developing a Go plugin
----------------------

Envoy's Golang HTTP-TCP Bridge plugin must implement the :repo:`HttpTcpBridge API <contrib/golang/common/go/api/filter.go>`.

Here is the introduction about the HttpTcpBridge API:

> - EncodeHeaders: get http request headers and decide whether to directly send rpc frame to tcp server.
> - EncodeData: get http request body and convert it to rpc frame, then send that to tcp server. and you can control whether to half-close upstream conn by Envoy.
> - OnUpstreamData: aggregate and verify multi rpc frame from tcp server, then convert complete rpc frame to http body, finally construct http response with headers to http client.

Reviewer 2: ```suggestion
- ``EncodeHeaders``: get HTTP request headers and decide whether to directly send RPC frame to TCP server.
- ``EncodeData``: get HTTP request body and convert it to RPC frame, then send that to TCP server. and you can control whether to half-close the upstream connection by Envoy.
- ``OnUpstreamData``: aggregate and verify multiple RPC frames from the TCP server, then convert the complete RPC frame to the HTTP body, finally construct the HTTP response with headers for the HTTP client.
```

Author: fix tit, thks~

.. attention::
  The Go plugin API is not yet stable, you are **strongly** recommended to use the same version of Go plugin SDK and Envoy.

When you are using a release version of Envoy, i.e. 1.26.x,
you should use ``github.com/envoyproxy/envoy v1.26.x`` in the go.mod file.

When you are not using a release, i.e. the latest main branch of Envoy,

> you could use ``go get -u github.com/envoyproxy/envoy@SHA`` to get the same version of Go plugin SDK,

Reviewer 2: ```suggestion
you could use ``go get -u github.com/envoyproxy/envoy@SHA`` to get the same version of the Go plugin SDK,
```

Author: fix tit, thks~

> the SHA is the latest commit sha.

Reviewer 2: ```suggestion
the SHA is the latest commit SHA.
```

Author: fix tit, thks~

Building a Go plugin
--------------------

.. attention::
  When building a Go plugin dynamic library, you **must** use a Go version consistent
  with Envoy's version of glibc.

One way to ensure a compatible Go version is to use the Go binary provided by Envoy's bazel setup:

.. code-block:: console

   $ bazel run @go_sdk//:bin/go -- version
   ...
   go version goX.YZ linux/amd64

For example, to build the ``.so`` for a ``foo`` plugin, you might run:

.. code-block:: console

   $ bazel run @go_sdk//:bin/go build -- --buildmode=c-shared  -v -o path/to/output/libfoo.so path/to/src/foo

Configuration
-------------

.. tip::
   This filter should be configured with the type URL
   :ref:`type.googleapis.com/envoy.extensions.upstreams.http.tcp.golang.v3alpha.Config <envoy_v3_api_msg_extensions.upstreams.http.tcp.golang.v3alpha.Config>`.

> A prebuilt Golang Http Tcp Bridge  ``my_bridge.so`` might be configured as follows:

Reviewer 2: ```suggestion
A prebuilt Golang HTTP TCP Bridge  ``my_bridge.so`` might be configured as follows:
```

Author: fix tit, thks~
